core/src/main/java/org/elasticsearch/search/profile/ProfileResult.java

```diff
@@ -162,4 +166,71 @@ public XContentBuilder toXContent(XContentBuilder builder, Params params) throws
+        XContentParser.Token token = parser.currentToken();
+        ensureExpectedToken(XContentParser.Token.START_OBJECT, token, parser::getTokenLocation);
+        String currentFieldName = null;
+        String type = null, description = null;
+        Map<String, Object> parsedTimings =  null;
+        List<ProfileResult> children = new ArrayList<>();
+        while((token = parser.nextToken()) != XContentParser.Token.END_OBJECT) {
+            if (token == XContentParser.Token.FIELD_NAME) {
+                currentFieldName = parser.currentName();
+            } else if (token.isValue()) {
+                if (TYPE.match(currentFieldName)) {
+                    type = parser.text();
+                } else if (DESCRIPTION.match(currentFieldName)) {
+                    description = parser.text();
+                } else if (NODE_TIME.match(currentFieldName)) {
+                    // we need to consume this value, but we use the raw nanosecond value
+                    parser.text();
+                } else if (NODE_TIME_RAW.match(currentFieldName)) {
+                    // the total time is calculate by adding up all timings internally, so we skip it
+                    parser.longValue();
+                } else {
+                    throwUnknownField(currentFieldName, parser.getTokenLocation());
+                }
+            } else if (token == XContentParser.Token.START_OBJECT) {
+                if (BREAKDOWN.match(currentFieldName)) {
+                    parsedTimings = parser.map();
+                } else {
+                    throwUnknownField(currentFieldName, parser.getTokenLocation());
+                }
+            } else if (token == XContentParser.Token.START_ARRAY) {
+                if (CHILDREN.match(currentFieldName)) {
+                    while ((token = parser.nextToken()) != XContentParser.Token.END_ARRAY) {
+                        children.add(ProfileResult.fromXContent(parser));
+                    }
+                } else {
+                    throwUnknownField(currentFieldName, parser.getTokenLocation());
+                }
+            }
+        }
+        Map<String, Long> timings = new HashMap<>(parsedTimings.size());
+        for (Entry<String, Object> entry : parsedTimings.entrySet()) {
+            Object value = entry.getValue();
+            if (value instanceof Integer) {
+                timings.put(entry.getKey(), ((Integer)value).longValue());
+            } else {
+                timings.put(entry.getKey(), (Long) value);
+            }
+        }
+        return new ProfileResult(type, description, timings, children);
+    }
+
+    /**
+     * Internal helper to calculate the time of a node, inclusive of children
+     *
+     * @param timings
+     *            A map of breakdown timing for the node
+     * @return The total time at this node, inclusive of children
+     */
+    private static long getNodeTime(Map<String, Long> timings) {
```

Comment:
Good point, I moved this code from somewhere where it was clear that the map is always there. I will add a check to the ctor.